src/interactions/panZoomInteraction.ts

```diff
@@ -161,13 +164,57 @@ export module Interactions {
 
         var deltaPixelAmount = e.deltaY * (e.deltaMode ? PanZoom._PIXELS_PER_LINE : 1);
         var zoomAmount = Math.pow(2, deltaPixelAmount * .002);
+
         this.xScales().forEach((xScale) => {
-          PanZoom._magnifyScale(xScale, zoomAmount, translatedP.x);
+          zoomAmount = this._constrainedZoomAmount(xScale, translatedP.x, zoomAmount);
         });
+
         this.yScales().forEach((yScale) => {
-          PanZoom._magnifyScale(yScale, zoomAmount, translatedP.y);
+          zoomAmount = this._constrainedZoomAmount(yScale, translatedP.y, zoomAmount);
+        });
+
+        this.xScales().forEach((xScale) => {
+          this._magnifyScale(xScale, zoomAmount, translatedP.x);
         });
+        this.yScales().forEach((yScale) => {
+          this._magnifyScale(yScale, zoomAmount, translatedP.y);
+        });
+      }
+    }
+
+    private _constrainedZoomAmount(scale: QuantitativeScale<any>, centerValue: number, zoomAmount: number) {
+      var extentIncreasing = zoomAmount > 1;
+
+      var boundingDomainExtent = extentIncreasing ? this.maxDomainExtent(scale) : this.minDomainExtent(scale);
+      if (boundingDomainExtent == null) { return zoomAmount; }
+
+      if (scale instanceof Scales.Linear || scale instanceof Scales.Time) {
+        var scaleDomain = (<any> scale).domain();
+        var domainExtent = Math.abs(scaleDomain[1] - scaleDomain[0]);
+        var compareF = extentIncreasing ? Math.min : Math.max;
+        return compareF(zoomAmount, boundingDomainExtent / domainExtent);
+      }
+
+      var constrainedZoomAmount = 1;
+      var lowerBound = extentIncreasing ? constrainedZoomAmount : 0;
+      var upperBound = extentIncreasing ? Infinity : constrainedZoomAmount;
+      var iterations = 20;
+      var magnifyTransform = (rangeValue: number) => scale.invert(centerValue - (centerValue - rangeValue) * constrainedZoomAmount);
+      var scaleRange = scale.range();
+      for (var i = 0; i < iterations; i++) {
+        var transformedDomain = scaleRange.map(magnifyTransform);
+        var transformedDomainExtent = Math.abs(transformedDomain[1] - transformedDomain[0]);
+        if (transformedDomainExtent === boundingDomainExtent) { return constrainedZoomAmount; }
+
+        if (extentIncreasing === transformedDomainExtent < boundingDomainExtent) {
```

Comment:
Can you add parens to this, for readers who don't know the [operator precedence rules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence) off the top of their head?